rfcs/0000_cortical_messaging_protocol_cmp.md

```diff
@@ -0,0 +1,153 @@
+- Start Date: 2025-08-06
+- RFC PR: (leave this empty, it will be filled in after RFC is merged)
+
+# Cortical Messaging Protocol (CMP)
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [BCP 14](https://www.rfc-editor.org/info/bcp14), [IETF RFC 2119](https://www.rfc-editor.org/rfc/rfc2119), [IETF RFC 8174](https://www.rfc-editor.org/rfc/rfc8174) when, and only when, they appear in all capitals, as shown here.
+
+## 1 Key Concepts
+
+CMP defines connectivity and communication within a Thousand Brains System.
+
+Communication within a Thousand Brains System is conveyed via variable-length Cortical Messages.
+
+CMP supports unicast and multicast delivery of data.
+
+The data flow in CMP is unidirectional, from the sender to one or more receivers.
+
+CMP is a connectionless protocol that maintains message boundaries with no connection setup or feature negotiation.
+
+### 1.1 Constraints
+
+The Thousand Brains System imposes a number of constraints that a CMP implementation MUST maintain.
+
+#### 1.1.1 Message Propagation Delay
+
+The Cortical Message propagation delay between the same sender and the same receiver MUST be constant for all steps.
+
+#### 1.1.2 Learning Modules Receive Before Vote
+
+Learning Modules MUST receive all Cortical Messages addressed to them before generating any Vote Cortical Messages for other Learning Modules.
```

Comment:
This currently only applies to messages that originate from SM or LM classification outputs (in CC this would be the L4/6 feed forward input). The LMs generate a vote output before they receive votes (which are also a CMP message). Those would be the L2/3 lateral connections in the CC.

Reply:
👍. If we retain this constraint, I'll clarify with an update to "... MUST receive all Observation Cortical Messages...".